Avoids crashes when broker is down and in ISR list

andrewjamesbrown opened the pull request:
In certain conditions, a partition ISR list includes a broker that Kazoo
does not currently have information about.  While this is rare, it is
exacerbated by restarting brokers and triggering topology changes.
There are also likely other (better) solutions to this; however this fix
is most expedient.

wvanbergen commented:
If we see an unknown broker ID, should we instead refresh the list of brokers from ZK?

andrewjamesbrown commented:
The issue is a race... in fact, we're refreshing the list of brokers from ZK _immediately before_ a call to `under_replicated?`.  The broker joins the cluster between those two calls, and hence can't be found when `set_state_from_json` looks for it.  I'm not keen on refreshing the list of brokers inside the partition class, since it mixes concerns too much.  At this point, we're implicitly refreshing the list ourselves in our caller, since it happens in a retry loop.

We can see a few options here:
- Swallow the error and return a smaller ISR list
- Raise a Kazoo exception (this is what we're doing in the PR)
- Teach Kazoo to use zookeeper watches in order to keep the broker list up-to-date.

@wvanbergen thoughts?

wvanbergen commented:
Option 3 is nice, but if you don't want to put in that work (understandable), I'd prefer an exception. At least that way people can build a retry mechanism on their own.

andrewjamesbrown commented:
@wvanbergen @sam-obeid : @psycotica0-shopify and I cleaned up the code (it now raises an exception) and added tests.  Looking for your +1 before merging.  Thanks!
